pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,224 @@
+		//master subnet
+		v, err := r.isSubnetNSGValid(ctx, masterSubnet, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg")
+		if err != nil {
+			return err
+		}
+
+		if !v {
+			err = r.fixSubnetNSG(ctx, masterSubnet, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg", instance.Spec.InfraID)
+			if err != nil {
+				return err
+			}
+		}
+		//worker subnets
+		for _, ws := range workerSubnets {
+			v, err := r.isSubnetNSGValid(ctx, ws, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg")
+			if err != nil {
+				return err
+			}
+			if !v {
+				err = r.fixSubnetNSG(ctx, ws, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg", instance.Spec.InfraID)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	default:
+		r.log.Errorf("Unknown value of Architecture Version %d", instance.Spec.ArchitectureVersion)
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) fixSubnetNSG(ctx context.Context, subnet string, vnet, resource azure.Resource, masterResourceGroup string, subnetsClient network.SubnetsClient, expectedName, infraID string) error {
+	fullSubnet, err := subnetsClient.Get(ctx, vnet.ResourceGroup, vnet.ResourceName, subnet, "")
+	if err != nil {
+		return err
+	}
+	fullSubnet.NetworkSecurityGroup = &mgmtnetwork.SecurityGroup{ID: to.StringPtr("/subscriptions/" + resource.SubscriptionID + "/resourceGroups/" + masterResourceGroup + "/providers/Microsoft.Network/networkSecurityGroups/" + expectedName)}
+	err = subnetsClient.CreateOrUpdateAndWait(ctx, vnet.ResourceGroup, vnet.ResourceName, subnet, fullSubnet)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func (r *AzureNSGReconciler) isSubnetNSGValid(ctx context.Context, subnet string, vnet, resource azure.Resource, masterResourceGroup string, subnetsClient network.SubnetsClient, expectedName string) (bool, error) {
+	fullSubnet, err := subnetsClient.Get(ctx, vnet.ResourceGroup, vnet.ResourceName, subnet, "")
+	if err != nil {
+		return false, err
+	}
+	if fullSubnet.SubnetPropertiesFormat == nil {
+		return false, fmt.Errorf("received nil SubnetProperties when trying to Get subnet %s/%s in resource group %s", vnet.ResourceName, subnet, vnet.ResourceGroup)
+	}
+	if fullSubnet.NetworkSecurityGroup == nil {
+		return false, fmt.Errorf("received nil NSG when trying to Get subnet %s/%s in resource group %s", vnet.ResourceName, subnet, vnet.ResourceGroup)
+	}
+	r.log.Printf("found NSG %s", *fullSubnet.NetworkSecurityGroup.ID)
+	nsg, err := azure.ParseResourceID(*fullSubnet.NetworkSecurityGroup.ID)
+	if err != nil {
+		return false, err
+	}
+	//subscription ID is the same as the cluster object, but resource group is the same as cluster resources (like master VMs)
+	return (nsg.SubscriptionID == resource.SubscriptionID && nsg.ResourceGroup == masterResourceGroup && (nsg.ResourceName == expectedName)), nil
```

Comment:
This has so much same code as dynamic validate, I bet we can reuse something? 
```
			nsgID, err := subnet.NetworkSecurityGroupID(oc, *ss.ID)
			if err != nil {
				return err
			}

			if ss.SubnetPropertiesFormat == nil ||
				ss.SubnetPropertiesFormat.NetworkSecurityGroup == nil ||
				!strings.EqualFold(*ss.SubnetPropertiesFormat.NetworkSecurityGroup.ID, nsgID) {
				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidLinkedVNet, s.Path, "The provided subnet '%s' is invalid: must have network security group '%s' attached.", s.ID, nsgID)
			}
```

Reply:
Changed the code to reuse the /pkg/util/subnet functions.